src/GraphQL/Resolver.hs

```diff
@@ -326,3 +336,123 @@ instance forall typeName interfaces fields m.
+    case extractUnionValue @(API.Object n i f) @union @m duv of
+      Just handler -> Just <$> (buildResolver @m @(API.Object n i f) handler selection)
+      Nothing -> runUnion @m @union @os duv fragment
+  runUnion _ _ = pure Nothing
+
+-- TODO(tom): AFAICT it should not be possible to ever hit the empty
+-- case because the compiler doesn't allow constructing a unionValue
+-- that's not in the Union. If the following code ever gets executed
+-- it's almost certainly a bug in the union code.
+--
+-- We still need to implement this instance for the compiler because
+-- it exhaustively checks all cases when deconstructs the Union.
+instance forall m union. RunUnion m union '[] where
+  runUnion _ _ = notImplemented
+
+instance forall m unionName objects.
+  ( Monad m
+  , KnownSymbol unionName
+  , RunUnion m (API.Union unionName objects) objects
+  ) => HasGraph m (API.Union unionName objects) where
+  type Handler m (API.Union unionName objects) = m (DynamicUnionValue (API.Union unionName objects) m)
+  buildResolver mHandler selectionSet = do
+    duv@(DynamicUnionValue _label _) <- mHandler
+    -- we only need to look at the fragment that matches by name:
+    case find (matchFragmentName _label) selectionSet of
+      Nothing -> pure (Result [UnionTypeNotFound] GValue.ValueNull) -- TODO more error detail
+      Just fragment -> do
+        -- loop through union handlers and call right one when type matches.
+        r <- runUnion @m @(API.Union unionName objects) @objects duv fragment
+        case r of
+          Just r' -> pure r'
+          Nothing -> pure (Result [ResolveNonUnionType] GValue.ValueNull)
+
+      where
+        matchFragmentName label' (SelectionInlineFragmentPattern (AST.NamedType name') _) =
+          label' == AST.getNameText name'
+        matchFragmentName _ _ = False
+
+
+symbolText :: forall ks. KnownSymbol ks => Text
+symbolText = toS (symbolVal @ks Proxy)
+
+-- | Translate a @Handler@ into a DynamicUnionValue type required by
+-- @Union@ handlers. This is dynamic, but nevertheless type-safe
+-- because we can only tag with types that are part of the union.
+--
+-- Use e.g. like "unionValue @Cat" if you have an object like this:
+--
+-- > type Cat = Object "Cat" '[] '[Field "name" Text]
+unionValue ::
+  forall (o :: Type) (union :: Type) m (name :: Symbol) i f.
+  (Monad m, API.Object name i f ~ o, KnownSymbol name)
+  => TypeIndex m o union -> m (DynamicUnionValue union m)
+unionValue x =
+  -- TODO(tom) - we might want to move to Typeable `cast` for uValue
+  -- instead of doing our own unsafeCoerce because it comes with
+  -- additional safety guarantees: Typerep is unforgeable, while we
+  -- can still into a bad place by matching on name only. We can't
+  -- actually segfault this because right now we walk the list of
+  -- objects in a union left-to-right so in case of duplicate names we
+  -- only every see one type. That doesn't seen like a great thing to
+  -- rely on though!
```

Comment:
Added another, more explicit note though I'm not super clear what you are looking for. There's also the comment on `DynamicUnionValue` which says the same thing in a different way.